### PR TITLE
Workaround for SWDEV-255735

### PR DESCRIPTION
**Pull request by shurale-nkn**
Now the assembler does a stronger check than before. Therefore, AMDHSA kernel xnack directive update are required.

`.amdhsa_reserve_xnack_mask` directive value (`1` or `0`) must match passed compiler argument (`-mxnack` or`-mno-xnack`) and curent GPU xnack mode (enabled or disabled). If the value does not match, the solver must be disabled.

Decided to disable xnack by default for asm kernels, since in most devices exclude APUs xnack disabled by default.

Full support for xnack requires the ability to detect the current GPU mode. This possibly can be done by using `Code Object Target Identification` which include xnack and ecc  mode ``amdgcn-amd-amdhsa--gfx906:sramecc-:xnack-``.

igemm kernels were updated using ``find ./ -type f -exec sed -i 's/.end_amdhsa_kernel/    \/\/xnack disabled by default for asm kernels\n    .amdhsa_reserve_xnack_mask 0\n.end_amdhsa_kernel/g' {} \;``

### CODE REVIEW

File: src/kernels/xform_bidirect_winograd_code.inc [hunk 1]
```diff
@@ -297,6 +297,7 @@ gid_x = 2
 .SGPR_ALLOC x_TW_stride
 .SGPR_ALLOC x_TH_stride
 .SGPR_ALLOC addr_step
+.SGPR_RESERVE_VCC
```
Comment (atamazov) [inline src/kernels/xform_bidirect_winograd_code.inc:300]: oops ;)

File: src/ocl/gcn_asm_utils.cpp [hunk 2]
```diff
@@ -239,7 +244,10 @@ static void AmdgcnAssembleQuiet(const std::string& source, const std::string& pa
 #ifdef __linux__
     std::stringstream clang_stdout_unused;
     const auto clang_path = GetGcnAssemblerPath();
-    const auto args       = " -x assembler -target amdgcn--amdhsa " + params + " " + source +
+    const auto args       = " -x assembler -target amdgcn--amdhsa " +
```
Comment (atamazov) [inline src/ocl/gcn_asm_utils.cpp:247]: Changes is in this file are not necessary. Why:
- Currently the library supports only parts that have XNACK mode OFF by default
  - See https://llvm.org/docs/AMDGPUUsage.html#processors
- The default value of the target feature is automatically set to the default XNACK setting of the target.
  - https://llvm.org/docs/AMDGPUUsage.html#target-features says: "_The target features supported by each processor, and the default value used if not specified explicitly, is listed in AMDGPU Processors._"

Moreover, this breaks compatibility with older assemblers that do not support the `-m[no-]xnack` option.

I am going to fix this right away.

**Review by atamazov — CHANGES_REQUESTED** [review]

File: src/kernels/Conv_Winograd_v13_3_12_epilogue.inc [hunk 3]
```diff
@@ -32,6 +32,11 @@ workgroup_size_x = 512
 .amdgcn.next_free_sgpr = 101
 .amdgcn.next_free_vgpr = 128
 
+//xnack disabled by default for asm kernels
```
Comment (atamazov) [inline src/kernels/Conv_Winograd_v13_3_12_epilogue.inc:35]: The changes like this are not necessary, but let's remove these later, when time permits.

Comment (atamazov) [reply src/kernels/Conv_Winograd_v13_3_12_epilogue.inc:35]: I mean that it's enough to change only one line.

File: src/kernels/dynamic_igemm/igemm_bwd_gtc_dynamic.s [hunk 4]
```diff
@@ -1227,6 +1227,8 @@ L_igemm_v4r1_bwd_dynamic_end:
     .amdhsa_next_free_sgpr 72
     .amdhsa_ieee_mode 0
     .amdhsa_dx10_clamp 0
+    //xnack disabled by default for asm kernels
+    .amdhsa_reserve_xnack_mask 0
```
Comment (atamazov) [inline src/kernels/dynamic_igemm/igemm_bwd_gtc_dynamic.s:1231]: [Future] Note that documentation says that the default of the `.amdhsa_reserve_xnack_mask` is Target Feature Specific, and, most likely, the changes like this are NOT required. Let's keep this change, however, for the sake of reliability. I think we will roll back such changes after upcoming release.

**Comment by atamazov** [comment]
It seems like `.amdhsa_reserve_xnack_mask 0` is not working without explicit `-mno-xnack` option. Another bug that requires a workaround. Working on it...

File: src/kernels/dynamic_igemm/igemm_bwd_gtc_dynamic.s (hunk 4 above)
Comment (shurale-nkn) [reply src/kernels/dynamic_igemm/igemm_bwd_gtc_dynamic.s:1231]: directive ``.amdhsa_reserve_xnack_mask``  added for each asm kernel at the request of management.

Comment (atamazov) [reply src/kernels/dynamic_igemm/igemm_bwd_gtc_dynamic.s:1231]: "management" is an abstract notion. Who?

**Review by atamazov — APPROVED** [review]
LGTM

File: src/ocl/gcn_asm_utils.cpp [hunk 5]
```diff
@@ -239,9 +261,9 @@ static void AmdgcnAssembleQuiet(const std::string& source, const std::string& pa
 #ifdef __linux__
     std::stringstream clang_stdout_unused;
     const auto clang_path = GetGcnAssemblerPath();
-    const auto args       = " -x assembler -target amdgcn--amdhsa " + params + " " + source +
-                      " -o /dev/null" + // We do not need output file
-                      " 2>&1";          // Keep console clean from error messages.
+    const auto args = " -x assembler -target amdgcn--amdhsa " + GetXnackOption() + params + " " +
```
Comment (atamazov) [inline src/ocl/gcn_asm_utils.cpp:264]: Bug is here. Suggestion: insert ` + " " ` after `GetXnackOption()`

**Pull request merged**